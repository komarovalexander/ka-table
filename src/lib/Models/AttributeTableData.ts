import { HTMLAttributes } from 'react';

import { ChildProps, DispatchFunc } from '../types';

export class AttributeTableData {
  public baseFunc!: any;
  public childElementAttributes!: HTMLAttributes<HTMLElement>;
  public childProps!: ChildProps;
  public dispatch!: DispatchFunc;
}
