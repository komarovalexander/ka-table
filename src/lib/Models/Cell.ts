/**
 * Describes the position of cell in table
 */
export class Cell {
  /*
  * Specifies the field of data
  */
  public columnKey!: string;

  /*
  * Specifies the value of row's key where cell is placed
  */
  public rowKey!: any;
}
