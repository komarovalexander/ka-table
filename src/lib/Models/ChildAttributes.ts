
import { ITableAllProps } from '../';
import { ICellContentProps } from '../Components/CellContent/CellContent';
import { ChildAttributesItem } from '../types';

export class ChildAttributes {
  public cell?: ChildAttributesItem<ICellContentProps>;
  public table?: ChildAttributesItem<ITableAllProps>;
}
