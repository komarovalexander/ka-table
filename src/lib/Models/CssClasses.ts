export class CssClasses {
  public table?: string = 'ka-table';
  public tbody?: string = 'ka-tbody';
  public thead?: string = 'ka-thead';
  public theadRow?: string = 'ka-tr ka-thead-row';
  public theadCell?: string = 'ka-thead-cell';
  public groupRow?: string = 'ka-tr ka-group-row';
  public row?: string = 'ka-tr ka-row';
  public rowSelected?: string = 'ka-row-selected';
  public cell?: string = 'ka-cell';

  public iconGroupArrowCollapsed?: string = 'ka-icon ka-icon-group-arrow ka-icon-group-arrow-collapsed';
  public iconGroupArrowExpanded?: string = 'ka-icon ka-icon-group-arrow ka-icon-group-arrow-expanded';
  public iconClose?: string = 'ka-icon ka-icon-close';
  public iconSortArrowDown?: string = 'ka-icon ka-icon-sort ka-icon-sort-arrow-down';
  public iconSortArrowUp?: string = 'ka-icon ka-icon-sort ka-icon-sort-arrow-up';

  public checkbox ?: string = 'ka-input';
  public dateInput ?: string = 'ka-input';
  public numberInput ?: string = 'ka-input';
  public textInput ?: string = 'ka-input';
}
