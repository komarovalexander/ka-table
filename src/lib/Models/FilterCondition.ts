/* Filter condition for specific field */
export class FilterCondition {
  public field!: string;
  public operator!: string;
  public value!: any;
}
