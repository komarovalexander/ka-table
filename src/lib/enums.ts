export enum DataType {
  Boolean = 'boolean',
  Date = 'date',
  Number = 'number',
  Object = 'object',
  String = 'string',
}

export enum EditingMode {
  None = 'none',
  Cell = 'cell',
}

export enum ActionType {
  CloseEditor = 'CloseEditor',
  ChangeFilterRow = 'ChangeFilterRow',
  OpenEditor = 'OpenEditor',
  ChangeRowData = 'ChangeRowData',
  DeselectRowData = 'DeselectRowData',
  SelectRow = 'SelectRow',
  ScrollTable = 'ScrollTable',
  ChangeSorting = 'ChangeSorting',
}

export enum KeyboardEnum {
  Esc = 27,
  Enter = 13,
}

export enum SortDirection {
  Ascend = 'ascend',
  Descend = 'descend',
}

export enum SortingMode {
  None = 'none',
  Single = 'single',
}

export enum FilteringMode {
  None = 'none',
  FilterRow = 'filterRow',
}

export enum FilterOperatorName {
  Equal = '=',
  MoreThan = '>',
  LessThan = '<',
  MoreThanOrEqual = '>=',
  LessThanOrEqual = '<=',
  Contains = 'contains',
}
