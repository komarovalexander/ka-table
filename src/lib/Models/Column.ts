import { DataType, SortDirection } from '../enums';
import {
  CellFunc, EditorFunc, FilterRowFunc, FormatFunc, HeaderCellFunc, SearchFunc, ValidationFunc,
} from '../types';

/**
 * Describes column of table its look and behaviour
 */
export class Column {
  public cell?: CellFunc;
  public dataType?: DataType;
  public editor?: EditorFunc;
  public filterRowCell?: FilterRowFunc;
  public filterRowValue?: any;
  public filterRowOperator?: any;
  public field?: string;
  public format?: FormatFunc;
  public headCell?: HeaderCellFunc;
  public isEditable?: boolean;
  public key!: string;
  public search?: SearchFunc;
  public sortDirection?: SortDirection;
  public style?: React.CSSProperties;
  public title?: string;
  public validation?: ValidationFunc;
}
