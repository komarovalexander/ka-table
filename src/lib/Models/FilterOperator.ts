import { DataType } from '../enums';

export class FilterOperator {
  public compare!: (fieldValue: any, conditionValue: any) => boolean;
  public defaultForTypes?: DataType[];
  public name!: string;
}
