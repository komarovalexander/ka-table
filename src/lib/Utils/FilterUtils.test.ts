import { DataType } from '../enums';
import { Column } from '../Models/Column';
import {
  filterData, getDefaultOperatorForType, getRowEditableCells, predefinedFilterOperators, searchData,
} from './FilterUtils';

describe('FilterUtils', () => {
  it('getRowEditableCells should return required cells from table EditableCells', () => {
    const rowEditableCells = getRowEditableCells(10, [{
      columnKey: '10',
      rowKey: 10,
    }, {
      columnKey: '10',
      rowKey: 2,
    }]);
    expect(rowEditableCells).toMatchSnapshot();
  });

  it('getDefaultOperatorForType', () => {
    expect(getDefaultOperatorForType(DataType.Number)).toBe('=');
    expect(getDefaultOperatorForType(DataType.Object)).toBe('=');
    expect(getDefaultOperatorForType(DataType.String)).toBe('contains');
  });

  describe('filterData', () => {
    const data: any[] = [
      { id: 1, name: 'Mike Wazowski', score: 80, passed: true },
      { id: 2, name: 'Billi Bob', score: 55, passed: false },
      { id: 3, name: 'Tom Williams', score: 45, passed: false },
      { id: 4, name: 'Kurt Cobain', score: 75, passed: true },
      { id: 5, name: 'Marshall Bruce', score: 77, passed: true },
      { id: 6, name: 'Sunny Fox', score: 33, passed: false },
    ];
    it('one item', () => {
      const columns = [{
        filterRowValue: 'Billi Bob',
        key: 'name',
      }];
      const result = filterData(data, columns);
      expect(result).toMatchSnapshot();
    });

    it('two item', () => {
      const columns = [{
        filterRowValue: 'Billi Bob',
        key: 'name',
      }, {
        dataType: DataType.Number,
        filterRowValue: 45,
        key: 'score',
      }];
      const result = filterData(data, columns);
      expect(result).toMatchSnapshot();
    });

    it('should throw an error in case of unknown filterOperator', () => {
      const columns = [{
        filterRowOperator: 'unknownOperator',
        filterRowValue: 'Billi Bob',
        key: 'name',
      }];
      expect(() => filterData(data, columns)).toThrowError('\'unknownOperator\' has not found in predefinedFilterOperators array, available operators: =, >, <, >=, <=, contains');
    });
  });

  [{
    falsy: [1, 2],
    name: '=',
    truthy: [1, 1],
  }, {
    falsy: [1, 1],
    name: '>',
    truthy: [2, 1],
  }, {
    falsy: [2, 1],
    name: '<',
    truthy: [1, 2],
  }, {
    falsy: [1, 2],
    name: '>=',
    truthy: [1, 1],
  }, {
    falsy: [2, 1],
    name: '<=',
    truthy: [1, 1],
  }, {
    falsy: ['abs', 'ss'],
    name: 'contains',
    truthy: ['hello', 'hell'],
  }].forEach((d) => {
    it(`predefinedFilterOperators operator ${d.name} truthy: ${d.truthy} falsy: ${d.falsy}`, () => {
      const operator = predefinedFilterOperators.find((o) => o.name === d.name);
      if (!operator) {
        throw new Error(`${d.name} was not found`);
      }
      expect(operator.compare(d.truthy[0], d.truthy[1])).toBeTruthy();
      expect(operator.compare(d.falsy[0], d.falsy[1])).toBeFalsy();
    });
  });

  describe('searchData', () => {
    const columns: Column[] = [
      { key: 'name', title: 'Name', dataType: DataType.String },
      { key: 'score', title: 'Score', dataType: DataType.Number },
      {
        dataType: DataType.Boolean,
        key: 'passed',
        search: (searchText?: string, rowData?: any) => {
          return (searchText === 'false' && !rowData.passed) || (searchText === 'true' && rowData.passed);
        },
        title: 'Passed',
      },
    ];
    const data: any[] = [
      { id: 1, name: 'Mike Wazowski', score: 80, passed: true },
      { id: 2, name: 'Billi Bob', score: 55, passed: false },
      { id: 3, name: 'Tom Williams', score: 45, passed: false },
      { id: 4, name: 'Kurt Cobain', score: 75, passed: true },
      { id: 5, name: 'Marshall Bruce', score: 77, passed: true },
      { id: 6, name: 'Sunny Fox', score: 33, passed: false },
    ];
    it('by string', () => {
      const result = searchData(columns, data, 'Mike');
      expect(result).toMatchSnapshot();
    });

    it('by boolean', () => {
      const columnsBoolean: Column[] = [
        {
          dataType: DataType.Boolean,
          key: 'passed',
          title: 'Passed',
        },
      ];
      const result = searchData(columnsBoolean, data, 'tru');
      expect(result).toMatchSnapshot();
    });

    it('should not find value by search handler', () => {
      const result = searchData(columns, data, 'tru');
      expect(result).toMatchSnapshot();
    });

    it('should find value by search handler', () => {
      const result = searchData(columns, data, 'true');
      expect(result).toMatchSnapshot();
    });
  });
});
