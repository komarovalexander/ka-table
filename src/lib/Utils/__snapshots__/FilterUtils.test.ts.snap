// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterUtils filterData one item 1`] = `
Array [
  Object {
    "id": 2,
    "name": "Billi Bob",
    "passed": false,
    "score": 55,
  },
]
`;

exports[`FilterUtils filterData two item 1`] = `Array []`;

exports[`FilterUtils getRowEditableCells should return required cells from table EditableCells 1`] = `
Array [
  Object {
    "columnKey": "10",
    "rowKey": 10,
  },
]
`;

exports[`FilterUtils searchData by boolean 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
  Object {
    "id": 4,
    "name": "Kurt Cobain",
    "passed": true,
    "score": 75,
  },
  Object {
    "id": 5,
    "name": "Marshall Bruce",
    "passed": true,
    "score": 77,
  },
]
`;

exports[`FilterUtils searchData by string 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
]
`;

exports[`FilterUtils searchData should find value by search handler 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
  Object {
    "id": 4,
    "name": "Kurt Cobain",
    "passed": true,
    "score": 75,
  },
  Object {
    "id": 5,
    "name": "Marshall Bruce",
    "passed": true,
    "score": 77,
  },
]
`;

exports[`FilterUtils searchData should not find value by search handler 1`] = `Array []`;
